rfc/0016-instances-vs-types/0016-instances-vs-types.md

```diff
@@ -0,0 +1,128 @@
+  - DISCUSSION: The RFC is open for discussion. Usually, we open a PR to trigger discussions.
+  - ACCEPTED: The RFC was accepted. Create issues to prepare implementation of the RFC.
+  - REJECTED: The RFC was rejected. If another revision emerges, switch to status DRAFT.
+-->
+
+## Summary
+Following the [Design Principles](https://jvalue.github.io/jayvee/docs/dev/design-principles), Jayvee focuses on describing a goal state instead of how to get there. That can lead to confusion about what is an object (instance of) vs. what is a type (definition of). 
+
+This RFC introduces a structure change that separates definitions from instantiation by moving definitions out of pipelines.
+
+## Motivation
+It is unclear what is an object (instance of) vs. what is a type (definition of) in Jayvee right now. With a small structure change, this confusion can be reduced while also providing positive side effects.
+
+## Explanation
+The following explanations use blocks as an example but map directly to **Constraints** and **Valuetypes**. That is, constraints and valuetypes can also only be **defined** outside of pipelines and are valuetypes are **instantiated** inside of pipelines (constraints are only instantiated as part of valuetypes).
+
+### Current state
+Pipelines can contain nearly any code (minimal cars example):
+```
+pipeline CarsPipeline {
+	CarsCSVExtractor 
+    -> CarsTableInterpreter
+		-> CarsLoader;
+
+	block CarsCSVExtractor oftype CSVExtractor {
+		url: "https://gist.githubusercontent.com/noamross/e5d3e859aa0c794be10b/raw/b999fb4425b54c63cab088c0ce2c0d6ce961a563/cars.csv";
+    enclosing: '"';
+	}
+	block CarsTableInterpreter oftype TableInterpreter {
+		header: true;
+		columns: [
+			"mpg" oftype decimal,
+			"cyl" oftype integer,
+			"disp" oftype decimal,
+		];
+	}
+	block CarsLoader oftype SQLiteLoader {
+		table: "Cars";
+		file: "./cars.sqlite";
+	}
+}
+```
+
+All block definitions (`block ... oftype ...`) are **definitions** of e.g., datasources or transformations (e.g. the `block CarsCSVExtractor oftype CSVExtractor` with all defined properties is a definition of the CSV file as a datasource).
+
+Instantiation happens only implicitly when a block reference is used in a pipeline, e.g. `CarsCSVExtractor -> CarsTableInterpreter` **instantiates** one instance of the `CarsCSVExtractor` during runtime. One block **definition** can be **instantiated** this way mutliple times in one pipeline.
+
+### Proposed change
+
+Allow **definitions** only outside of pipelines, allow block **instantiations** only inside of pipelines.
+
+Because pipelines get executed implicitly when executing a Jayvee model (and therefore pipelines get **instantiated** during runtime), it makes sense to bundle all other instantiations in them as well. This means everything outside of a pipeline is a **definition** (**type**), everything inside a pipeline is an **instance**.
```

Comment:
Do we always instantiate as singletons within a pipeline?

Reply:
I don't think so but I am also not sure I understand what you are asking. The interpreter instantiates a new object for every block reference in a pipeline I think.

Reply:
I think it is a special kind of instantiation: you cannot create to instances of a block, right?
I don't know how to better frame it >.<

Reply:
Do you mean in Jayvee or in the interpreter?

Reply:
I think in Jayvee as well. Let's take this example:
```
pipeline Demonstrator {
    MyCsvExtractor
      -> MyTableInterpreter
      -> MySqliteLoader;
 
  MyTableInterpreter
    -> MyPostgresLoader; 
}
```

In my understanding, one instance of `MyTableInterpreter` would be shared between those two branches of the graph.

Reply:
I don't think that is true, we instantiate a new executor for every block reference we find (see `createBlockExecutor` in `block-executor-registry.ts` and where it is called, e.g. in `runPipeline`). It would be easy to change that or add some cache but right now we don't (as far as I can see in my very short check :D).

Reply:
Now I'm confused :D I thought that we would execute `MyTableInterpreter` in this example only once and cache the result data for the later block in the execution order.. But I could be wrong :)

Reply:
Ah sorry yeah, in your example that is probably right. Because we traverse the pipeline in topological order so it would be `MyCsvExtractor`, then `MyTableInterpreter` and then the child blocks. That does not mean we would create a global singleton though, if the block is used in a different pipeline it would get executed again (?).

Reply:
Exactly, but within the scope of a pipeline it kinda is. The question is if that is just an interpreter detail or part of the language as well..

Reply:
However, that is not always true. Example:

```
pipeline MySecondPipeline {
  MyExtractor1 -> MyTableInterpreter -> MyLoader;

  MyExtractor2 -> MyTableInterpreter -> MyLoader;
}
```

Reply:
Sorry, I think I lost your suggestion on how to change this RFC for this. I think this is an implementation detail of the interpreter so I am unsure what effect it should have here. Can you make that more explicit? 😅

Reply:
Okay, my suggestion is to specify when an instance is "reuse" like in the first example and when not like in the second example. However, I'm having troubles to specify this. Maybe leave it open then?

Reply:
Ah, I would leave this out of the scope of this to focus on the description vs. instantiation semantics. Reusing existing instances on instantiation is a nuance that would make this too big I think?